### PR TITLE
Refactored code and created tensorflow backend

### PR DESCRIPTION
**Pull request by Thenerdstation**
This is a big change, but I think it is for the better. (See https://github.com/google/TensorNetwork/issues/51 for details). 

The next backend we will add will be a pure numpy implementation, then JAX, and then leave torch to the community.

### CODE REVIEW

**Comment by Thenerdstation** [comment]
This is a big design change so I would appreciate your inputs.

File: tensornetwork/backends/base_backend.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    Args: 
+      tensor: A tensor.
+    Returns:
+      The reshaped tensor.
+    """
+    raise NotImplementedError(
+        "Backend '{}' has not implemented reshape.".format(self.name))
+
+  def transpose(self, tensor: Tensor, perm: Sequence[int]) -> Tensor:
+    """Transpose a tensor to a given permutation.
```
Comment (sleichen) [inline tensornetwork/backends/base_backend.py]: I would say "Transpose a tensor according to a given permutation"

File: tensornetwork/backends/base_backend.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      "Backend '{}' has not implemented concat.".format(
+          self.name))
+
+  def shape(self, tensor: Tensor) -> Tensor:
+    """Get the shape of a tensor.
+
+    Args:
+      tensor: A tensor.
+    Returns:
+      The shape of the tensor as a different tensor.
```
Comment (sleichen) [inline tensornetwork/backends/base_backend.py]: How about "The shape of the input tensor returned as another tensor."

File: tensornetwork/backends/base_backend.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  def sqrt(self, tensor: Tensor) -> Tensor:
+    """Take the square root (element wise) of a given tensor."""
+    raise NotImplementedError(
+      "Backend '{}' has not implemented sqrt.".format(
+          self.name))
+
+  def diag(self, tensor: Tensor) -> Tensor:
+    """Create a diagonal matrix from the given vector tensor."""
+    raise NotImplementedError(
+      "Backend '{}' has not implemented sqrt.".format(
```
Comment (sleichen) [inline tensornetwork/backends/base_backend.py]: sqrt --> diag

File: tensornetwork/backends/base_backend.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  def trace(self, tensor: Tensor) -> Tensor:
+    """Calculate the trace over the last two axes of the given tensor."""
+    raise NotImplementedError(
+      "Backend '{}' has not implemented trace.".format(
+          self.name))
+
+  def outer_product(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
+    """Calculate the outer product of the two given tensors."""
+    raise NotImplementedError(
+      "Backend '{}' has not implemented trace.".format(
```
Comment (sleichen) [inline tensornetwork/backends/base_backend.py]: trace --> outer_product

**Comment by Thenerdstation** [comment]
> Also, I'm not sure how switching backends will be implemented.

Two ways (I'm waiting on stavros building the numpy backend before trying it). 
1. Doing `tensornetwork.TensorNetwork(backend='numpy')`
2. Defaulting to the global backend, and changing the global backend with `tensornetwork.set_backend('numpy')`

> but it seems conceivable that someone could try to merge two networks created using two different backends. How do we deal with that?

Easiest thing to do is throw an error. We can check that the backends are the same before merging, which is pretty easy. We *could* have a converter by changing each node's tensor by doing `torch.Tensor -> np.array -> tf.Tensor`, but I expect that would get messy and I'm not sure people would use it often enought to justify the complexity. 

All 4 backends we plan on supporting right now (JAX/PyTorch/TF/numpy) allow tensors to be created by converting a `np.array` to their own `Tensor` type. All higher level network creators we build would just have to build nodes using numpy and they would get converted internally.

File: tensornetwork/backends/base_backend.py (hunk 4 above)
Comment (Thenerdstation) [reply tensornetwork/backends/base_backend.py]: Can you tell I copy and pasted? lol

**Comment by Thenerdstation** [comment]
> network_components and ncon_interface need to be updated as well, right?

Yes they do,

**Pull request merged**